File: 2024/Day02.cpp
#include <iostream>
#include <map>
#include <string>
#include <vector>

#include "utils.hpp"

#include "Days.hpp"

Day02::Day02(string path) : AdventDay(path) {
    day = 2;
    title = "Red-Nosed Reports";
}

long Day02::isGoodReport(const vector<long> report) {
    int dir = 0;
    for(unsigned int i = 1; i < report.size(); i++) {
      long delta = report[i] - report[i-1];

      if(delta==0 || abs(delta) > 3) {
        // bad step
        return 0;
      }

      int currentDir = delta<0 ? -1 : 1;
      if(dir==0) {
        dir = currentDir;
      } else if(dir != currentDir) {
        // change of direction
        return 0;
      }
    }
    return 1;
}

void Day02::part1(vector<string> lines) {
  long safeReport = 0;

  for(auto line : lines) {
    vector<long> report = extractLongs(line);
    safeReport += isGoodReport(report);
  }

  printLongResult(safeReport);
}

void Day02::part2(vector<string> lines) {
  long safeReport = 0;

  for(auto line : lines) {
    vector<long> report = extractLongs(line);

    long current = isGoodReport(report);

    if(current == 1) {
      safeReport++;
    } else {
      for(unsigned int i = 0; i < report.size(); i++) {
        vector<long> reportCopy(report);
        reportCopy.erase(reportCopy.begin() + i);
        if(isGoodReport(reportCopy)) {
          safeReport++;
          break;
        }
      }
    }
  }

  printLongResult(safeReport);
}

File: 2024/python/day02.py
from aoc import *

def isSafe(report):
    direction = 0
    for i in range(1, len(report)):
        delta = report[i] - report[i-1]
        if delta==0 or abs(delta)>3:
            return 0

        currentdir = 1 if delta>0 else -1
        if direction==0:
            direction = currentdir
        else:
            if currentdir != direction:
                return 0
    return 1     

def part1(lines):
    res = 0
    for l in lines:
        report = [int(x) for x in l.split()]
        res += isSafe(report)
    return res

def part2(lines):
    res = 0
    for l in lines:
        report = [int(x) for x in l.split()]
        if isSafe(report):
            res += 1
        else:
            for i in range(len(report)):
                if isSafe(report[:i] + report[i+1:]):
                    res += 1
                    break
    return res

AoCRunnerAll(2, 'Red-Nosed Reports', part1, part2)